Allow user to add names to addresses.

MatthewKennedy opened the pull request:
Allows user to add an optional, but unique address name to each of their address.

### HTML CSS Improvements

- Moved the edit and delete address icons outside of the html label, no nested buttons in a selectable element.
- Added red border to fields with submit errors giving clear indication of errors.
- Now text says **New Address** and **Edit Address** was **New Shipping Address** and **Edit Shipping Address**.
### Tests Strengthened 
- User FFaker for randomised address data
- Check the cloned user_id to be_nil
- Check the cloned label to be_nil (Address names need to be unique, not cloned).

MatthewKennedy committed:
Allow usr to add names to addresses.

damianlegawiec commented on api/app/helpers/spree/api/api_helpers.rb:
```diff
@@ -116,7 +116,7 @@ def required_fields_for(model)
         :id, :firstname, :lastname, :full_name, :address1, :address2, :city,
         :zipcode, :phone, :company, :alternative_phone, :country_id, :state_id,
-        :state_name, :state_text
+        :location_name, :state_name, :state_text
```
@MatthewKennedy could we call this new field `label`? That's probably the most universal one

MatthewKennedy replied:
I’ll make this change.

damianlegawiec commented on core/app/models/spree/address.rb:
```diff
@@ -30,6 +30,11 @@ class Address < Spree::Base
     end
 
     validate :state_validate, :postal_code_validate
+    validates_uniqueness_of :location_name,
```
`validates_uniqueness_of` helper is deprecated, we should use `validates :location_name, uniqueness: {`

MatthewKennedy replied:
And this change also

MatthewKennedy committed:
Update validation

MatthewKennedy committed:
Update address.rb

MatthewKennedy committed:
Make requested changes

MatthewKennedy committed:
Fix Code Climate issues

damianlegawiec commented on frontend/app/assets/javascripts/spree/frontend/checkout/address.js:
```diff
@@ -42,7 +42,6 @@ Spree.ready(function ($) {
           zipcodeLabel.text(zipText)
           zipcodeInput.addClass('required')
         } else {
-          zipcodeInput.val('')
```
@MatthewKennedy is this required?

MatthewKennedy replied:
This just causes a small issue where by if someone has entered a zip code although it’s not required for their country, if they the. go back to edit the address on page load it blanks out the zip code.

If you were to have the zip code field removed from the DOM then that would be needed.

MatthewKennedy replied:
This looks like its been merged from the other PR

damianlegawiec commented on core/spec/models/spree/address_spec.rb:
```diff
@@ -32,6 +32,7 @@
       expect(cloned.state_id).to eq(original.state_id)
       expect(cloned.state_name).to eq(original.state_name)
       expect(cloned.zipcode).to eq(original.zipcode)
+      expect(cloned.user_id).to eq(original.user_id)
```
@MatthewKennedy actually, there is catch to this - we're cloning addresses from user to order when the user completes the checkout. Those cloned addresses shouldn't be associated with the user anymore - there should be only associated to order

MatthewKennedy replied:
We may need a different solution then.

If you edit an address that’s not editable because it’s tied to an existing order, the clone happens to keep the address for the completed order, but upon save the newly cloned address is not accessible to the user.

Video of this prop to this change

https://youtu.be/a5GbSMAYHbs

MatthewKennedy replied:
@damianlegawiec I've been looking at this over the past few hours, and as far as I can tell, the address should be associated with a user after checkout so that the address they just entered at checkout is then available in the user accounts section.

The clone attributed that I have changed in this PR, comes into play if the user then tries to edit an address that is already associated with an existing order, at this point the address from the order is cloned and presented to the user as an edit, even though technically its a new address.

This preserves the address id and details from the original address for the order details.

This video shows both the bug and the fix in action, don't know if it helps.
https://youtu.be/ix3HB7La4Rs

MatthewKennedy replied:
@damianlegawiec If you like, I'll pull the change out of this PR?

MatthewKennedy replied:
I've pulled this change out as it can be dealt with in its own PR.

MatthewKennedy committed:
Use FFaker for clone testing, and remove unused css.

MatthewKennedy committed:
Remove cloning of user_id

MatthewKennedy committed:
CSS nicer, ppinter on delete icon, hover darnens color.

MatthewKennedy committed:
Merge branch 'master' into address-names

MatthewKennedy committed:
Remove location svg, leave in the currently unused info.svg?

MatthewKennedy committed:
Stronger address tests.

damianlegawiec committed:
Merge branch 'master' into address-names

MatthewKennedy commented:
@damianlegawiec - Do you want me to look into this PR and find why the test is failing? If I remember correctly, it was a nice addition to address management.

damianlegawiec commented:
Hey @MatthewKennedy it would be great to merge this, it's a really welcome addition we see demand for in many projects

MatthewKennedy committed:
Update HTML.

MatthewKennedy commented:
@damianlegawiec That should be okay now.

MatthewKennedy committed:
Remove duplicate style from css, and remove unused class on edit.svg

MatthewKennedy commented:
I'm just going to see if I can put some of the manual css I added during the workflow into the existing scss,  tidy things up a bit.

MatthewKennedy committed:
Move address book reference from CSS to imported through SCSS to get access to variables.

MatthewKennedy committed:
Remove hard coded color.

MatthewKennedy committed:
Add spacing

MatthewKennedy committed:
Don't need to give a class to the SVG anymore.

MatthewKennedy committed:
Remove unnecessary SVG class.

MatthewKennedy committed:
Use nice edit icon form admin.

MatthewKennedy committed:
Refine the address edit / delete buttons.

MatthewKennedy committed:
Put distance between clickable elements.

MatthewKennedy commented:
That should be better. Let me know if changes are needed after another review.

I moved a css file reference from the main css manifest imported via the application.scss to pick up variables.

Also removed the require self, do you want that putting back in the css manifest file?

MatthewKennedy committed:
Add back in require self.

damianlegawiec commented:
Thanks @MatthewKennedy 👍 This is good to go!
